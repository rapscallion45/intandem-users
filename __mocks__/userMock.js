const accountMock = {
  displayName: 'Jaydon Frankie',
  firstName: 'Jaydon',
  lastName: 'Frankie',
  username: 'jaydon.frankie',
  email: 'jaydon.frankie@test.com',
  photoURL: '/static/mock-images/avatars/avatar_default.jpg',
  password: 'testtest',
};

export default accountMock;
