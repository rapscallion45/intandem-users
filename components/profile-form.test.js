import React from 'react';
import { shallow } from 'enzyme';
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import toJson from 'enzyme-to-json';
import userMock from '../__mocks__/userMock';
import ProfileForm from './profile-form';

describe('Profile Form', () => {
  it('Renders correctly enzyme', () => {
    /* Arrange */
    const mockCancel = jest.fn();
    const mockSave = jest.fn();
    const mockSaving = false;
    const mockShowCancel = true;

    /* Act */
    const wrapper = shallow(
      <ProfileForm
        userData={userMock}
        handleCancel={mockCancel}
        handleSave={mockSave}
        saving={mockSaving}
        showCancel={mockShowCancel}
      />
    );

    /* Assert */
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe('Form Fields', () => {
    it('Should render the correct input fields', () => {
      /* Arrange */
      const mockCancel = jest.fn();
      const mockSave = jest.fn();
      const mockSaving = false;
      const mockShowCancel = true;

      /* Act */
      const wrapper = render(
        <ProfileForm
          userData={userMock}
          handleCancel={mockCancel}
          handleSave={mockSave}
          saving={mockSaving}
          showCancel={mockShowCancel}
        />
      );

      /* Assert */
      expect(wrapper.getByLabelText('Email')).toBeInTheDocument();
      expect(wrapper.getByLabelText('First Name')).toBeInTheDocument();
      expect(wrapper.getByLabelText('Last Name')).toBeInTheDocument();
    });
  });

  describe('Form Operation', () => {
    it('Should render Save User button', async () => {
      /* Arrange */
      const mockCancel = jest.fn();
      const mockSave = jest.fn();
      const mockSaving = false;
      const mockShowCancel = true;
      const saveBtnText = 'Save';
      const user = userEvent.setup();

      /* Act */
      const wrapper = render(
        <ProfileForm
          userData={userMock}
          handleCancel={mockCancel}
          handleSave={mockSave}
          saving={mockSaving}
          showCancel={mockShowCancel}
        />
      );
      await user.click(wrapper.getByText(saveBtnText));

      /* Assert */
      expect(wrapper.queryByText(saveBtnText)).toBeInTheDocument();
      await waitFor(() => {
        expect(mockSave).toHaveBeenCalledTimes(1);
      });
    });

    it('Should not allow Save button functionality if form not complete', async () => {
      /* Arrange */
      const mockCancel = jest.fn();
      const mockSave = jest.fn();
      const mockSaving = false;
      const mockShowCancel = true;
      const saveBtnText = 'Save';

      /* Act */
      const wrapper = render(
        <ProfileForm
          userData={userMock}
          handleCancel={mockCancel}
          handleSave={mockSave}
          saving={mockSaving}
          showCancel={mockShowCancel}
        />
      );
      act(() => {
        fireEvent.change(wrapper.getByLabelText('Last Name'), { target: { value: '' } });
      });
      fireEvent.click(wrapper.getByText(saveBtnText));

      /* Assert */
      expect(wrapper.queryByText(saveBtnText)).toBeInTheDocument();
      expect(mockSave).toHaveBeenCalledTimes(0);
    });

    it('Should render Cancel button if prop set', () => {
      /* Arrange */
      const mockCancel = jest.fn();
      const mockSave = jest.fn();
      const mockSaving = false;
      const mockShowCancel = true;
      const cancelBtnText = 'Cancel';

      /* Act */
      const wrapper = render(
        <ProfileForm
          userData={userMock}
          handleCancel={mockCancel}
          handleSave={mockSave}
          saving={mockSaving}
          showCancel={mockShowCancel}
        />
      );
      fireEvent.click(wrapper.getByText(cancelBtnText));

      /* Assert */
      expect(wrapper.queryByText(cancelBtnText)).toBeInTheDocument();
      expect(mockCancel).toHaveBeenCalledTimes(1);
    });

    it('Should not render Cancel button if prop not set', () => {
      /* Arrange */
      const mockCancel = jest.fn();
      const mockSave = jest.fn();
      const mockSaving = false;
      const mockShowCancel = false;
      const cancelBtnText = 'Cancel';

      /* Act */
      const wrapper = render(
        <ProfileForm
          userData={userMock}
          handleCancel={mockCancel}
          handleSave={mockSave}
          saving={mockSaving}
          showCancel={mockShowCancel}
        />
      );

      /* Assert */
      expect(wrapper.queryByText(cancelBtnText)).toBeNull();
    });

    it('Should render spinner in place of save button when saving profile', () => {
      /* Arrange */
      const mockCancel = jest.fn();
      const mockSave = jest.fn();
      const mockSaving = true;
      const mockShowCancel = true;
      const saveBtnText = 'Save';

      /* Act */
      const wrapper = render(
        <ProfileForm
          userData={userMock}
          handleCancel={mockCancel}
          handleSave={mockSave}
          saving={mockSaving}
          showCancel={mockShowCancel}
        />
      );

      /* Assert */
      expect(wrapper.queryByText(saveBtnText)).toBeNull();
      expect(wrapper.queryByTestId('save-spinner')).toBeInTheDocument();
    });
  });
});
