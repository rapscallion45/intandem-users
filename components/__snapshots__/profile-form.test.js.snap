// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Form Renders correctly enzyme 1`] = `
<form
  onSubmit={[Function]}
>
  <ForwardRef(Grid)
    container={true}
    spacing={2}
  >
    <ForwardRef(Grid)
      item={true}
      xs={12}
    >
      <ForwardRef(TextField)
        fullWidth={true}
        id="first_name"
        label="First Name"
        margin="normal"
        name="first_name"
        onChange={[Function]}
        type="first_name"
        value="Michael"
        variant="outlined"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      item={true}
      xs={12}
    >
      <ForwardRef(TextField)
        fullWidth={true}
        id="last_name"
        label="Last Name"
        margin="normal"
        name="last_name"
        onChange={[Function]}
        type="last_name"
        value="Lawson"
        variant="outlined"
      />
    </ForwardRef(Grid)>
    <ForwardRef(Grid)
      item={true}
      xs={12}
    >
      <ForwardRef(TextField)
        fullWidth={true}
        id="email"
        label="Email"
        margin="normal"
        name="email"
        onChange={[Function]}
        type="email"
        value="michael.lawson@reqres.in"
        variant="outlined"
      />
    </ForwardRef(Grid)>
  </ForwardRef(Grid)>
  <ForwardRef(Grid)
    item={true}
    xs={12}
  >
    <ForwardRef(Grid)
      container={true}
      justifyContent="end"
      mt={1}
      spacing={2}
    >
      <ForwardRef(Grid)
        item={true}
        xs={12}
      >
        <ForwardRef(Button)
          color="success"
          disabled={false}
          fullWidth={true}
          type="submit"
          variant="contained"
        >
          Save
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
      <ForwardRef(Grid)
        item={true}
        xs={12}
      >
        <ForwardRef(Button)
          fullWidth={true}
          onClick={[MockFunction]}
          variant="contained"
        >
          Cancel
        </ForwardRef(Button)>
      </ForwardRef(Grid)>
    </ForwardRef(Grid)>
  </ForwardRef(Grid)>
</form>
`;
